File: python3/perfect_squares.py
"""
Space   : O(n)
Time    : O(n**2)
"""


class Solution:
    def numSquares(self, n: int) -> int:
        dp = [0] + [None] * n

        for i in range(1, n + 1):
            j = 1
            min_val = 10000000

            while j * j <= i:
                min_val = min(min_val, dp[i - j * j])
                j += 1

            dp[i] = min_val + 1

        return dp[-1]

File: golang/perfect_squares.go
package main

/*
Space	: O(n)
Time	: O(n log n)
*/

func numSquares(n int) int {
	dp := make([]int, n+1)

	for i := 1; i <= n; i++ {
		dp[i] = 1000
	}

	for i := 0; i <= n; i++ {
		j := 0

		for i-j*j >= 0 {
			sq := j * j
			dp[i] = Min(dp[i], dp[i-sq]+1)
			j++
		}
	}

	return dp[n]
}
